software/controller/lib/core/nvparams.cpp

```diff
@@ -0,0 +1,235 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This module handles non-volatile parameter storage
+//
+// These parameters are stored in a dedicated 256 kbits flash memory.
+//
+// On system startup we read through the flash to initialize the parameters.
+//
+// A checksum and flip/flop system ensure the integrity of the parameters: write
+// operations write the data to one side of the flip/flop, update the checksum,
+// (which validates that side), and then update the data on the other side (but
+// not its checksum, which renders it invalid) in preparation of the next write.
+
+#include "nvparams.h"
+#include "checksum.h"
+#include "eeprom.h"
+#include "hal.h"
+#include "vars.h"
+#include <string.h>
+
+// Size of the parameter block including the header
+static constexpr uint32_t nvparam_size{sizeof(NVparams_struct)};
+
+// local functions
+static bool ReadFullParams(NVParamsAddress address, void *param);
+static void WriteFullParams(NVParamsAddress address, void *param);
+static uint32_t ParamsCRC(void *param);
+static bool IsValid(NVparams_struct *param) {
```

Comment:
I think this could also be defined in the header and be testable? This also seems like a candidate for method?

Reply:
@martukas does it make sense to have it in the public interface? what would be a use case of someone calling it? It's always better to keep public interfaces and header files generally as small as possible imo